src/capability/mod.rs

```diff
@@ -0,0 +1,134 @@
+    CompactPCICentralResourceControl(PciCapabilityAddress),
+    /// PCI Standard Hot-Plug Controller capability, Cap ID = `0x0C`
+    PciHotPlugControl(PciCapabilityAddress),
+    /// Bridge subsystem vendor/device ID capability, Cap ID = `0x0D`
+    BridgeSubsystemVendorId(PciCapabilityAddress),
+    /// AGP Target PCI-PCI bridge capability, Cap ID = `0x0E`
+    AGP3(PciCapabilityAddress),
+    /// PCI Express capability, Cap ID = `0x10`
+    PciExpress(PciCapabilityAddress),
+    /// MSI-X capability, Cap ID = `0x11`
+    MsiX(PciCapabilityAddress),
+    /// Unknown capability
+    Unknown {
+        address: PciCapabilityAddress,
+        id: u8,
+    },
+}
+
+impl PciCapability {
+    fn parse(id: u8, address: PciCapabilityAddress, control: u16) -> Option<PciCapability> {
+        match id {
+            0x00 => None, // null capability
+            0x01 => Some(PciCapability::PowerManagement(address)),
+            0x02 => Some(PciCapability::AcceleratedGraphicsPort(address)),
+            0x03 => Some(PciCapability::VitalProductData(address)),
+            0x04 => Some(PciCapability::SlotIdentification(address)),
+            0x05 => Some(PciCapability::Msi(MsiCapability::new(address, control))),
+            0x06 => Some(PciCapability::CompactPCIHotswap(address)),
+            0x07 => Some(PciCapability::PciX(address)),
+            0x08 => Some(PciCapability::HyperTransport(address)),
+            0x09 => Some(PciCapability::Vendor(address)),
+            0x0A => Some(PciCapability::DebugPort(address)),
+            0x0B => Some(PciCapability::CompactPCICentralResourceControl(address)),
+            0x0C => Some(PciCapability::PciHotPlugControl(address)),
+            0x0D => Some(PciCapability::BridgeSubsystemVendorId(address)),
+            0x0E => Some(PciCapability::AGP3(address)),
+            0x10 => Some(PciCapability::PciExpress(address)),
+            0x11 => Some(PciCapability::MsiX(address)),
+            _ => Some(PciCapability::Unknown { address, id }),
+        }
+    }
+}
+
+pub struct CapabilityIterator<'a, T: ConfigRegionAccess> {
+    address: PciAddress,
+    offset: u16,
+    access: &'a T,
+}
+
+impl<'a, T: ConfigRegionAccess> CapabilityIterator<'a, T> {
+    pub(crate) fn new(
+        address: PciAddress,
+        offset: u16,
+        access: &'a T,
+    ) -> CapabilityIterator<'a, T> {
+        CapabilityIterator {
+            address,
+            offset,
+            access,
+        }
+    }
+}
+
+impl<'a, T: ConfigRegionAccess> Iterator for CapabilityIterator<'a, T> {
+    type Item = PciCapability;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            if self.offset == 0 {
+                return None;
+            }
+            let data = unsafe { self.access.read(self.address, self.offset) };
+            let next_ptr = data.get_bits(8..16);
+            let id = data.get_bits(0..8);
+            let control = data.get_bits(16..32) as u16;
```

Comment:
Hmm so is included with some but not all of the capabilities, and you just know based on the type? Maybe a name like `extension` would be better?

Reply:
It is included with all PCI capabilities, but have different meaning depending on capability type. And yes, `extension` will be a better name than `control`.